pkg/planner/core/find_best_task.go

```diff
@@ -1172,13 +1172,15 @@ func skylinePruning(ds *logicalop.DataSource, prop *property.PhysicalProperty) [
 		}
 	}
 
-	preferRange := ds.SCtx().GetSessionVars().GetAllowPreferRangeScan() && (ds.TableStats.HistColl.Pseudo || ds.TableStats.RowCount < 1)
 	// If we've forced an index merge - we want to keep these plans
 	preferMerge := len(ds.IndexMergeHints) > 0 || fixcontrol.GetBoolWithDefault(
 		ds.SCtx().GetSessionVars().GetOptimizerFixControlMap(),
 		fixcontrol.Fix52869,
 		false,
 	)
+	// tidb_opt_prefer_range_scan is the master switch to control index preferencing
+	preferRange := ds.SCtx().GetSessionVars().GetAllowPreferRangeScan() &&
+		(preferMerge || (ds.TableStats.HistColl.Pseudo || ds.TableStats.RowCount < 1))
```

Comment:
i don't get why should we put preferMerge into this equation,seems irrelevant? it seems change the orignal logic, may be this is reason why irrelevant sql plan result is changed, or you are trying to fix something

Reply:
It seems like we can treat `IndexMerge` as a special kind of `IndexRange`, so `preferMerge` also means the user prefers some `IndexRange`, which allow us to set `preferRange=true` here.